app/src/main/java/com/github/raulccabreu/redukt/ui/StateView.kt

```diff
@@ -0,0 +1,49 @@
+package com.github.raulccabreu.redukt.ui
+
+import android.content.Context
+import android.view.View
+import com.github.raulccabreu.redukt.Redukt
+import com.github.raulccabreu.redukt.states.StateListener
+
+abstract class StateView<T>(context: Context) : View(context), StateListener<T> {
+
+    private var isRegistered = false
+
+    protected abstract fun changeStateWhenHidden(): Boolean
+    protected abstract fun getRedukt(): Redukt<T>
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+
+        registerReducer()
+    }
+
+    override fun onDetachedFromWindow() {
+        unregisterReducer()
+        super.onDetachedFromWindow()
+    }
+
+    override fun setVisibility(visibility: Int) {
+        super.setVisibility(visibility)
+        if (visibility == View.VISIBLE && !isRegistered)
+            registerReducer()
+        else if (visibility != View.VISIBLE && !changeStateWhenHidden())
+            unregisterReducer()
+    }
+
+    private fun registerReducer() {
```

Comment:
Hmm! We aren't register it in a reducer, it is registering itself to hear state changes. Thus I think that a more appropriated name could be, registerStateChanges for example or other related.